### ISSUE TITLE
DOC404 issue with Generator

### ISSUE THREAD

**Issue by oliversheridanmethven** [jsh9/pydoclint#279]
Running pydoclint 0.8.3 with Python 3.14 I still see this error: https://github.com/jsh9/pydoclint/issues/198

Specifically with the example:

```
"""Generator Test."""

from __future__ import annotations

from typing import TYPE_CHECKING


if TYPE_CHECKING:
    from collections.abc import Generator


def yield_5() -> Generator[int]:
    """
    Docstring test.

    Yields
    ------
    int
        5.
    """
    yield 5
```

I still see `Function `yield_5` yield type(s) in docstring not consistent with the return annotation. The yield type (the 0th arg in Generator[...]/Iterator[...]): Generator[int]; docstring "yields" section types: int`. 

However, in this PR https://github.com/jsh9/pydoclint/pull/199 it says it fixed the issue. I don't think the fix has persisted and the original issue was closed prematurely.

**Comment by jsh9** [comment]
Hi,

In such cases, it's technically incorrect to use `Generator[int]` as the type annotator.  The correct alternatives are:

* `Iterator[int]`
* `Generator[int, ??, ??]` (fill in actual types to the two `??` placeholders)

Here's AI's explanation:

---

### Why `Generator[int]` is considered "incorrect"

If you only provide one argument, Python's type checkers (like Mypy or Pyright) can get confused. They don't always "default" to `None` for the missing pieces; instead, they might treat the missing arguments as `Any`, which weakens your type safety.

To be explicitly correct for your specific code, you should write:

```python
def yield_5() -> Generator[int, None, None]:
    """Yields 5 and nothing else."""
    yield 5

```

### A Cleaner Alternative: `Iterator`

If you find the `Generator[int, None, None]` syntax too verbose (and most people do), there is a more idiomatic way to handle it.

If your generator only yields values and doesn't use `.send()` or a final `return` value, you can use **`Iterator`** or **`Iterable`** instead. These only require a single type argument:

```python
from collections.abc import Iterator

def yield_5() -> Iterator[int]:
    yield 5

```

> **Tip:** Use `Iterator[int]` for 90% of your generator functions. Only break out the full `Generator[Yield, Send, Return]` when you are actually doing advanced coroutine-style programming with `.send()`.

---

I'll make a code change to add the explanation above into the violation message of DOC404.

**Comment by cebtenzzre** [comment]
@jsh9 The explanation you've quoted from AI is incorrect in a couple of ways.

**`Generator[int]` is valid Python 3.13 syntax.** The abbreviated form `Generator[YieldType]` (and `Generator[YieldType, SendType]`) was made explicit in Python 3.13 — the stdlib typeshed stubs for `Generator` define the `SendType` and `ReturnType` parameters with defaults of `None`, meaning `Generator[int]` is precisely equivalent to `Generator[int, None, None]`. This is not a type-checker quirk; it's a first-class language feature. Both mypy and pyright handle it correctly with no errors or warnings.

**The "might treat missing arguments as `Any`" claim is false.** That would only be true if the parameters had no defaults. They do. You can verify this in the [typeshed stub for `Generator`](https://github.com/python/typeshed/blob/54bea668cb7c976500c1e1f890351c8cdaabd080/stdlib/typing.pyi#L559-L564).

**The irony here is that basedpyright actively *warns* against `Iterator` on `@contextmanager` functions** — it emits `reportDeprecated` if you annotate a `@contextmanager`-decorated function as `-> Iterator[Foo]`, telling you to use `-> Generator[Foo]` instead. So for at least one major type checker, the "cleaner alternative" recommended here is the option that triggers a deprecation warning.

This is a pydoclint bug — specifically, DOC404 not recognising the abbreviated `Generator[X]` form as equivalent to `Generator[X, None, None]`. You acknowledged this yourself in [#198](https://github.com/jsh9/pydoclint/issues/198): *"I recently learnt that starting from Python 3.13, `Generator[int]` and `Generator[int, None]` can be considered equivalent to `Generator[int, None, None]`, but I haven't got a chance to implement this yet."*

The correct fix is to implement that handling in pydoclint.

**Comment by jsh9** [comment]
OK.

If either of you have time, please feel free to submit a PR and I'll review it.  Otherwise, I'll fix it when I have free time.  (No guarantee when that will be though because I'm tied up in other projects at the moment.)

Thanks!

**Comment by jsh9** [comment]
For future reference, it is [PEP 696](https://peps.python.org/pep-0696/) that established the new syntax to allow abbreviated `Generator[X]` type annotation.
